vent/coordinator/coordinator.py

```diff
@@ -0,0 +1,156 @@
+import time
+import threading
+from typing import List
+
+from vent.common.message import SensorValues, ControlSetting, Alarm, ControlSettingName, IPCMessageCommand
+from vent.controller.control_module import get_control_module
+from vent.coordinator.ipc import IPC
+from vent.coordinator.process_manager import ProcessManager
+
+
+class CoordinatorBase:
+    def __init__(self, sim_mode=False):
+        # get_ui_control_module handles single_process flag
+        self.control_module = get_control_module(sim_mode)
+        self.sensor_values = None
+        self.alarms = None
+        self.control_settings = {}
+        self.tentative_control_settings = {}
+        self.last_message_timestamp = None
+
+    def get_sensors(self) -> SensorValues:
+        # returns SensorValues struct
+        pass
+
+    # def get_alarms(self) -> List[Alarm]:
+    #     # returns list of Alarm structs
+    #     pass
+
+    def get_active_alarms(self) -> List[Alarm]:
+        pass
+
+    def get_logged_alarms(self) -> List[Alarm]:
+        pass
+
+    def clear_logged_alarms(self):
+        pass
+
+    def set_control(self, control_setting: ControlSetting):
+        # takes ControlSetting struct
+        pass
+
+    def get_control(self, control_setting_name: ControlSettingName) -> ControlSetting:
+        pass
+
+    def get_msg_timestamp(self):
+        # return timestamp of last message
+        pass
+
+    def do_process(self, command: IPCMessageCommand):
+        # TODO: we need to test these
+        # start / stop / reset process
+        if command == IPCMessageCommand.START:
+            self.control_module.start()
+        elif command == IPCMessageCommand.STOP:
+            self.control_module.stop()
+        else:
+            raise KeyError("Error: undefined action" + str(command))
+
+
+class CoordinatorLocal(CoordinatorBase):
+    def __init__(self, sim_mode=False):
+        super().__init__(sim_mode=sim_mode)
+        self.lock = threading.Lock()
+        self.thread = threading.Thread(target=self.start, daemon=True)
+        self.thread.start()
+        self.thread_id = self.thread.ident
+
+    def __del__(self):
+        self.thread.join()
+
+    def set_control(self, control_setting: ControlSetting):
+        self.lock.acquire()
+        self.tentative_control_settings[control_setting.name] = control_setting
+        self.lock.release()
+
+    def get_control(self, control_setting_name: ControlSettingName) -> ControlSetting:
+        self.lock.acquire()
+        control_setting = self.control_settings[control_setting_name]
```

Comment:
Do we need to make a copy of control_setting? Is it returning a pointer or a copy?

Reply:
My intention is to make a copy. This is because `self.control_settings` is shared by `start()` thread and `get_control()` function. By making a copy I can safely return without worry being modified by `start()` background.

However, I realize I made a mistake here. This line actually returns a pointer.

Reply:
Sounds good.